### PR TITLE
[Feature][Authentication] support oauth2 login

### PR DESCRIPTION
**Pull request by hdygxsj**
## Purpose of the pull request

support oauth2 login
and close issue #14056

## How users use oauth2

First create an Oauth2 app in github(Users can also use their own OAuth2 authorization server, this is just an example)
image

Configure dolphinscheduler to support OAuth2
b4e7ed4528bc626f950a6912ed007ae

A button for oauth2 login will appear on the login interface after enabling oauth2
36a766de365af872cc725ad6ab7c839

Click the github button to jump to the authorization page
8618c5153f84b67e5ca1907a7cdbc8a

Jump to Homepage after authorization

775e0b6615ed662eeb86b69f88c8a88

## Brief change log

front end：
* The login page will get the oauth2 providers
* Clicking the button of oauth2 will jump to the authorization page and clear the current cookie

backend：
* Added an api to clear cookies
* Added an api to get OAuth2 providers
* Added an api that gets access_token and user information and redirects to the home page
In this api, after the user successfully authorizes using oauth2, if the user does not exist, a user will be created automatically
* Added a Configuration for automatic assembly oauth2 configuration

### CODE REVIEW

**Comment by hdygxsj** [comment]
This pr does not change the original login method, nor does it introduce spring security, it just adds a few interfaces and a little change to support oauth2, can you take a look when you are free @EricGao888

File: dolphinscheduler-ui/src/views/login/index.tsx [hunk 1]
```diff
@@ -38,15 +41,16 @@ import { useLocalesStore } from '@/store/locales/locales'
 import { useThemeStore } from '@/store/theme/theme'
 import cookies from 'js-cookie'
 import { ssoLoginUrl } from '@/service/modules/login'
+import { OAuth2Provider } from '@/service/modules/login/types'
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/index.tsx]: Please import type.

File: dolphinscheduler-ui/src/views/login/index.tsx [hunk 2]
```diff
@@ -38,15 +41,16 @@ import { useLocalesStore } from '@/store/locales/locales'
...
+
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/index.tsx]: Please delete enter.

File: dolphinscheduler-ui/src/views/login/index.tsx [hunk 3]
```diff
@@ -170,6 +178,15 @@ const login = defineComponent({
               </NButton>
             </a>
           </div>
+            {this.oauth2Providers.length > 0 ? <NDivider >
+              {this.t('login.otherwayLogin')}
+            </NDivider> : <div></div>}
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/index.tsx]: Don't use Ternary operator, Please user this.oauth2Providers.length > 0 && <NDivider>
              {this.t('login.otherwayLogin')}
            </NDivider>

File: dolphinscheduler-ui/src/views/login/use-login.ts [hunk 4]
```diff
@@ -16,23 +16,25 @@
...
 import type { Router } from 'vue-router'
-import type { LoginRes } from '@/service/modules/login/types'
+import { useRoute } from 'vue-router'
+import type { LoginRes, OAuth2Provider } from '@/service/modules/login/types'
 import type { UserInfoRes } from '@/service/modules/users/types'
 import { useRouteStore } from '@/store/route/route'
 import { useTimezoneStore } from '@/store/timezone/timezone'
 import cookies from 'js-cookie'
 import { queryBaseDir } from '@/service/modules/resources'
+import { reactive, ref } from 'vue'
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/use-login.ts]: Please remove unused references.

File: dolphinscheduler-ui/src/views/login/use-login.ts [hunk 5]
```diff
@@ -63,7 +65,49 @@ export function useLogin(state: any) {
...
+  const oauth2Providers = ref<Array<OAuth2Provider> | []>([])
+
+  const gotoOAuth2Page = async (oauth2Provider: OAuth2Provider) => {
+    await clearCookie()
+    window.location.href = `${oauth2Provider.authorizationUri}?client_id=${oauth2Provider.clientId}` +
+      `&redirect_uri=${oauth2Provider.redirectUri}?provider=${oauth2Provider.provider}`
+  }
+
+  const handleRedirect = async () => {
+    debugger
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/use-login.ts]: Please remove debugger.

File: dolphinscheduler-ui/src/views/login/use-login.ts [hunk 6]
```diff
@@ -63,7 +65,49 @@ export function useLogin(state: any) {
...
+    const authType = route.query.authType
+    if (authType && authType == 'oauth2') {
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/use-login.ts]: Please use '==='.

File: dolphinscheduler-ui/src/views/login/use-login.ts [hunk 7]
```diff
@@ -63,7 +65,49 @@ export function useLogin(state: any) {
...
+        await userStore.setUserInfo(userInfoRes)
+        const timezone = userInfoRes.timeZone ? userInfoRes.timeZone : 'UTC'
+        await timezoneStore.setTimezone(timezone)
+        router.push('home')
+      }
+      const error = route.query.error
+      if (error) {
+        window.$message.error(error)
+      }
+
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/use-login.ts]: Please delete enter.

File: dolphinscheduler-ui/src/views/login/use-login.ts [hunk 8]
```diff
@@ -63,7 +65,49 @@ export function useLogin(state: any) {
...
+    }
+
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/login/use-login.ts]: Please delete enter.

**Review by songjianet — CHANGES_REQUESTED** [review]

File: dolphinscheduler-ui/src/views/login/index.tsx (hunk 1 above)
Comment (hdygxsj) [reply dolphinscheduler-ui/src/views/login/index.tsx]: I have already resolve it, could you please help review it again when you are free @songjianet

**Review by qingwli — COMMENTED** [review]
Could you update the doc?

**Comment by hdygxsj** [comment]
> Could you update the doc?

Of course, I'll finish it

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java [hunk 9]
```diff
@@ -160,4 +180,77 @@ public Result signOut(@Parameter(hidden = true) @RequestAttribute(value = Consta
         request.removeAttribute(Constants.SESSION_USER);
         return success();
     }
+
+    @DeleteMapping("cookies")
+    public void clearCookieSessionId(HttpServletRequest request, HttpServletResponse response) {
+        Cookie[] cookies = request.getCookies();
+        for (Cookie cookie : cookies) {
+            cookie.setMaxAge(0);
```
Comment (EricGao888) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java:188]: Just for double check, do we need to set the cookie path here?

File: dolphinscheduler-ui/src/locales/en_US/login.ts [hunk 10]
```diff
@@ -22,5 +22,6 @@ export default {
   userPassword: 'Password',
   userPassword_tips: 'Please enter your password',
   login: 'Login',
+  otherwayLogin: 'login by otherways',
```
Comment (EricGao888) [inline dolphinscheduler-ui/src/locales/en_US/login.ts]: What about giving the tips as `OAuth2.0`? It may be more intuitive to users.

**Comment by EricGao888** [comment]
@hdygxsj BTW, we had `Casdoor` login which is also based on `OAuth`. If we could find some way to make it configurable as one of the login providers in the future, it would be nice. However, it is out of the scope of this PR, you don't need to do it here.

https://github.com/apache/dolphinscheduler/pull/8706

**Comment by EricGao888** [comment]
Backend part basically LGTM, if related docs added.

**Review by songjianet — DISMISSED** [review]

**Comment by hdygxsj** [comment]
> Backend part basically LGTM, if related docs added.

I found something wrong with the backend redirection api that I still need to fix

**Comment by EricGao888** [comment]
> > Backend part basically LGTM, if related docs added.
> 
> I found something wrong with the backend redirection api that I still need to fix

May I ask whether it is possible to cover the `OAuth` authentication process in our API / E2E test cases in the future?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java (hunk 9 above)
Comment (hdygxsj) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java:188]: Yes, now with OAuth2 authorization, the back end will create the user using the user information and return the sessionId.  However, if a user tries to login as a local user before login with OAuth2, two Sessionids will be generated in the cookie, and the Sessionids generated after OAuth2 authorization cannot overwrite the sessionIds generated, resulting in an authorization failure.To sum up, before jumping to the OAuth2 authorization page, we need to clear the current cookie.

**Comment by hdygxsj** [comment]
> 

Yes, but we need to create a github oauth2 application and create an oauth2 client credence, or we can start an authorization server ourselves in an end-to-end test. I can try.

**Comment by hdygxsj** [comment]
> > > Backend part basically LGTM, if related docs added.
> > 
> > 
> > I found something wrong with the backend redirection api that I still need to fix
> 
> May I ask whether it is possible to cover the `OAuth` authentication process in our API / E2E test cases in the future?


Yes, but we need to create a github oauth2 application and create an oauth2 client credence, or we can start an authorization server ourselves in an end-to-end test. I can try.

**Comment by hdygxsj** [comment]
> Backend part basically LGTM, if related docs added.

Document has been updated

**Comment by Radeity** [comment]
> > Backend part basically LGTM, if related docs added.
> 
> Document has been updated

Hi @hdygxsj , the content of Chinese and English doc should better be consistent.

**Comment by hdygxsj** [comment]
> > > Backend part basically LGTM, if related docs added.
> > 
> > 
> > Document has been updated
> 
> Hi @hdygxsj , the content of Chinese and English doc should better be consistent.

Have been updated

**Comment by EricGao888** [comment]
> > > > Backend part basically LGTM, if related docs added.
> > > 
> > > 
> > > Document has been updated
> > 
> > 
> > Hi @hdygxsj , the content of Chinese and English doc should better be consistent.
> 
> Have been updated

I just reran the CI, waiting for it to pass.

**Comment by hdygxsj** [comment]
截屏2023-08-17 19 58 12

Is there anything I need to fix about this error in k8s e2e test

**Comment by Radeity** [comment]
> Is there anything I need to fix about this error in k8s e2e test

No need, it has already been fixed in other pr.

File: docs/docs/en/guide/security/authentication-type.md [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # Authentication Type
 
-* So far we support three authentication types, Apache DolphinScheduler password, LDAP and Casdoor SSO.
+* So far we support three authentication types, Apache DolphinScheduler password, LDAP,Casdoor SSO and OAuth2，the OAuth2 authorization login mode can be used with other authentication modes.
```
Comment (qingwli) [inline docs/docs/en/guide/security/authentication-type.md]:
```suggestion
* So far we support three authentication types, Apache DolphinScheduler password, LDAP, Casdoor SSO and OAuth2，the OAuth2 authorization login mode can be used with other authentication modes.
```

File: docs/docs/zh/guide/security/authentication-type.md [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # 认证方式
 
-* 目前我们支持三种认证方式，Apache DolphinScheduler自身账号密码登录，LDAP和通过Casdoor实现的SSO登录。
+* 目前我们支持三种认证方式，Apache DolphinScheduler自身账号密码登录，LDAP,通过Casdoor实现的SSO登录和通过Oauth2授权登录，并且oauth2授权登录方式可以和其他认证方式同时使用。
```
Comment (qingwli) [inline docs/docs/zh/guide/security/authentication-type.md]:
```suggestion
* 目前我们支持三种认证方式，Apache DolphinScheduler自身账号密码登录，LDAP, 通过Casdoor实现的SSO登录和通过Oauth2授权登录，并且Oauth2授权登录方式可以和其他认证方式同时使用。
```

File: docs/docs/en/guide/security/authentication-type.md (hunk 11 above)
Comment (hdygxsj) [reply docs/docs/en/guide/security/authentication-type.md]: done

**Review by EricGao888 — APPROVED** [review]

**Review by qingwli — APPROVED** [review]
+1

**Comment by EricGao888** [comment]
@hdygxsj This is a nice feature. Thanks for your contributions : )

**Comment by yonghen1** [comment]
Hello, I would like to ask if it is possible to integrate the dolphinscheduler with its own internal oauth2, and if you can provide me with a complete configuration file, I am still having a problem with oauth2 auth error, what is going wrong
微信图片_20240813143305

**Comment by hdygxsj** [comment]
> Hello, I would like to ask if it is possible to integrate the dolphinscheduler with its own internal oauth2, and if you can provide me with a complete configuration file, I am still having a problem with oauth2 auth error, what is going wrong 微信图片_20240813143305

Currently, there may be compatibility issues when obtaining tokens and user information, and I will attempt to refactor this feature.

**Comment by yonghen1** [comment]
@hdygxsj The problem of oauth2 auth error seems to be the error message returned by the code. Is this caused by compatibility? Is there any other way to integrate your own internal oauth2 protocol

**Pull request merged**